File: frontend/src/api/credits.ts
import { api } from '@/api/client'

export type CreditKind = 'signup' | 'recharge' | 'gift' | 'consume' | 'refund'
export type PackId = 'starter' | 'standard' | 'pro'

export type CreditCosts = {
  generate: number
  edit: number
  upscale: number
  marketing: number
}

export type CreditPack = { id: PackId; label: string; credits: number }

export type CreditCatalog = {
  balance: number
  costs: CreditCosts
  packs: CreditPack[]
}

export type LedgerEntry = {
  id: string
  amount: number
  balance_after: number
  kind: CreditKind
  reason: string
  created_at: string
}

export const KIND_LABELS: Record<CreditKind, string> = {
  signup: '注册赠送',
  recharge: '充值',
  gift: '管理员赠送',
  consume: '消耗',
  refund: '退回',
}

export const creditsApi = {
  catalog: () => api.get<CreditCatalog>('/credits/catalog'),
  ledger: () => api.get<LedgerEntry[]>('/credits/ledger'),
  recharge: (pack_id: PackId) => api.post<CreditCatalog>('/credits/recharge', { pack_id }),
}

File: backend/app/schemas/credits.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.models.credit import CreditKind, CreditLedger

PackId = Literal["starter", "standard", "pro"]


class CreditCostsOut(BaseModel):
    generate: int
    edit: int
    upscale: int
    marketing: int


class CreditPackOut(BaseModel):
    id: PackId
    label: str
    credits: int


class CreditCatalogOut(BaseModel):
    balance: int
    costs: CreditCostsOut
    packs: list[CreditPackOut]


class RechargeIn(BaseModel):
    pack_id: PackId


class LedgerOut(BaseModel):
    id: uuid.UUID
    amount: int
    balance_after: int
    kind: CreditKind
    reason: str
    created_at: datetime

    @classmethod
    def of(cls, entry: CreditLedger) -> "LedgerOut":
        return cls(
            id=entry.id,
            amount=entry.amount,
            balance_after=entry.balance_after,
            kind=entry.kind,
            reason=entry.reason,
            created_at=entry.created_at,
        )


class GiftIn(BaseModel):
    amount: int = Field(gt=0, le=10_000)
    reason: str = Field(default="管理员赠送", max_length=120)

    @field_validator("reason")
    @classmethod
    def _normalize_reason(cls, value: str) -> str:
        return value.strip() or "管理员赠送"
